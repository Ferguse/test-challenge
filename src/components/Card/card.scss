@import "../../styles/variables.scss";

.card {
  padding: 30px 50px 30px 10px;
  color: var(--category-color);

  @media screen and (max-width: 600px) {
    padding: 15px;
  }

  @media screen and (max-width: 480px) {
    padding: 10px;
  }

  &-content {
    transition: $transition;
    box-shadow: 2px 2px 10px 1px rgba(0, 0, 0, 0.2);
    min-height: 320px;

    &:hover {
      box-shadow: 2px 2px 17px 5px rgba(0, 0, 0, 0.2);
    }
  }

  &-image {
    padding: 10px;
    position: relative;
    height: 200px;
  }

  &-info {
    height: 120px;
    padding: 10px;
    text-align: center;
  }

  &-icon {
    margin: 0;
    position: absolute;
    top: 50%;
    transform: translate(0, -50%);
    color: $lightblue3;
    z-index: -1;
  }
}
