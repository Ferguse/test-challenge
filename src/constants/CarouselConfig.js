export const carouselConfig = {
  dots: true,
  infinite: true,
  speed: 500,
  slidesToShow: 5,
  slidesToScroll: 5,
  responsive: [
    {
      breakpoint: 1024,
      settings: {
        slidesToShow: 3,
        slidesToScroll: 3,
        arrows: false,
      },
    },
    {
      breakpoint: 600,
      settings: {
        slidesToShow: 2.5,
        slidesToScroll: 2,
        initialSlide: 1,
        arrows: false,
      },
    },
    {
      breakpoint: 480,
      settings: {
        slidesToShow: 1.3,
        slidesToScroll: 1,
        arrows: false,
      },
    },
  ],
};
