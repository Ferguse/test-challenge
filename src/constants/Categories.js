export const CATEGORIES = [
  { name: 'bike', icon: 'bike', color: '#ffa2a2', font: 'SourceSans' },
  { name: 'hiking', icon: 'hiking', color: '#a7fda7', font: 'Swansea' },
  { name: 'skating', icon: 'skating', color: '#a9a9fb', font: 'Goldleaf' },
];
