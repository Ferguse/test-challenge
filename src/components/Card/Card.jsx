import { Link } from 'react-router-dom';
import { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';

import { transformOverflowText } from 'utils/transformOverflowText';
import { setCurrentBook, setCurrentCategory } from 'actions';

import './card.scss';

const Card = ({ id, volumeInfo, category }) => {
  const dispatch = useDispatch();
  const style = { '--category-color': category.color };

  const handleClick = useCallback(() => {
    dispatch(setCurrentBook(id));
    dispatch(setCurrentCategory(category.name));
  }, [id, category]);

  return (
    <div className="card" style={style}>
      <Link to={`/category/${category.name}/book/${id}`} onClick={handleClick}>
        <div className="card-content flex direction-column justify-between">
          <div className="card-image flex justify-center">
            <img src={volumeInfo?.imageLinks?.thumbnail} alt={volumeInfo.title} loading="lazy"/>
            <i className="fi fi-rs-books card-icon"></i>
          </div>
          <div className="card-info flex direction-column">
            <span className="card-title bold">{transformOverflowText(volumeInfo.title)}</span>
            <span>{new Date(volumeInfo.publishedDate).getFullYear()}</span>
          </div>
        </div>
      </Link>
    </div>
  );
};

Card.propTypes = {
  id: PropTypes.string.isRequired,
  volumeInfo: PropTypes.shape({
    title: PropTypes.string.isRequired,
    imageLinks: PropTypes.shape({
      thumbnail: PropTypes.string,
    }),
    publishedDate: PropTypes.string,
  }).isRequired,
  category: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }).isRequired,
};

export { Card };
