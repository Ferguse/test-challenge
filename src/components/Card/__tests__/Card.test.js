import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockedStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import { CATEGORIES } from 'constants/Categories';
import { Card } from '../Card';

describe('components [Card]', () => {
  const state = {
    wishList: {},
    categories: {
      entities: [],
    },
    current: {},
  };

  const props = {
    category: CATEGORIES[0],
    volumeInfo: {
      publishedDate: '2020',
      title: 'test',
    },
    id: 'id',
  };

  const store = configureMockedStore([thunk])(state);
  const history = createMemoryHistory({ initialEntries: ['/'] });

  it('Renders', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router history={history} location={history.location} navigator={history}>
            <Card {...props} />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
