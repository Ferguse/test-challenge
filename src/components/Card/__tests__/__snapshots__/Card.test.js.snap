// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components [Card] Renders 1`] = `
<div
  className="card"
  style={
    {
      "--category-color": "#ffa2a2",
    }
  }
>
  <a
    href="/category/bike/book/id"
    onClick={[Function]}
  >
    <div
      className="card-content flex direction-column justify-between"
    >
      <div
        className="card-image flex justify-center"
      >
        <img
          alt="test"
          loading="lazy"
        />
        <i
          className="fi fi-rs-books card-icon"
        />
      </div>
      <div
        className="card-info flex direction-column"
      >
        <span
          className="card-title bold"
        >
          test
        </span>
        <span>
          2020
        </span>
      </div>
    </div>
  </a>
</div>
`;
